### PR TITLE
MRG: refactor propspeed

### PR DESCRIPTION
**Pull request by matthew-brett**
Refactoring propspeed to make me feel better about not understanding or being
able to debug strange error on buildbots.

### CODE REVIEW

**Comment by arokem** [comment]
Does this solve that bug on that *one* buildbot? Otherwise, this refactor should probably wait for 0.9

We'll need to review this, and it's not a trivial amount of work. We should focus on putting out 0.8 first.

**Comment by Garyfallidis** [comment]
I second Ariel's post. Also if you are planning on putting that much amount effort you may consider that the new tracking API made by Bago and Gab should replace EuDX in the future. So, it might be wiser to focus on optimizing and improving that.

**Comment by matthew-brett** [comment]
Guys - the effort is done.  Now I am going to add some checks on top.  That will be done soon too.

There's nothing much to review here, just docstring edits and code formatting.   Maybe y'all could save time by having a quick look now to confirm this is an improvement.

**Comment by matthew-brett** [comment]
OK - refactoring done.  I think it's uncontroversial, and improves the code, making it a little safer against numpy changes.

**Comment by matthew-brett** [comment]
I will also get travis to build a wheel with these changes on top of the 0.7.1 release to the resulting wheel fixes the buildbot error.

**Comment by arokem** [comment]
Wait - does this fix that error on your machine? Under 3.3? That was not
clear to me from your previous messages

**Comment by matthew-brett** [comment]
I built the 0.7.1 wheel with the patches from this PR, and failing test does now pass.  Wheel build here: https://travis-ci.org/matthew-brett/dipy-wheels/builds/46025087  .  Wheels at http://wheels.scipy.org/ .  I propose to upload the fixed (patched) wheel for Python 3.3 to pypi.

**Comment by arokem** [comment]
I don't understand - you want to upload this instead of 0.7.1? It's
completely different

**Comment by arokem** [comment]
Sorry - that was a brainfart on my part. More power to you!

**Comment by arokem** [comment]
I guess the answer to my question is a yes? This does fix that bug? In that case, I will review this later tonight.

**Comment by matthew-brett** [comment]
Yes, if I rebase the changes in this PR on top of 0.7.1 - and rebuild a wheel with that code, then the test failure is fixed for Python 3.3.   The wheel is then the same as for 0.7.1 apart from these changes.   Actual branch I built the wheel from is : https://github.com/matthew-brett/dipy/tree/tmp  .  So yes, this appears to fix a bug.

**Comment by matthew-brett** [comment]
When reviewing, maybe review by commit, because all but the last commit are docstring / formatting cleanups.

File: dipy/tracking/propspeed.pyx [hunk 1]
```diff
@@ -89,73 +88,104 @@ def ndarray_offset(cnp.ndarray[cnp.npy_intp, ndim=1] indices, \
...
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def map_coordinates_trilinear_iso(cnp.ndarray[double, ndim=3] data,\
-                                   cnp.ndarray[double, ndim=2] points,\
-                                   cnp.ndarray[cnp.npy_intp, ndim=1] data_strides,\
-                                   cnp.npy_intp len_points,\
-                                   cnp.ndarray[double, ndim=1] result):
-    ''' trilinear interpolation (isotropic voxel size)
+def map_coordinates_trilinear_iso(
+    cnp.ndarray[double, ndim=3] data,
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: We usually align parameters below the first parameter.

Comment (matthew-brett) [inline dipy/tracking/propspeed.pyx]: ​That would cause a line longer than 78 characters, hence the change.

File: dipy/tracking/propspeed.pyx [hunk 2]
```diff
@@ -89,73 +88,104 @@ def ndarray_offset(cnp.ndarray[cnp.npy_intp, ndim=1] indices, \
...
+    if not cnp.PyArray_CHKFLAGS(data_strides, cnp.NPY_C_CONTIGUOUS):
+        raise ValueError(u"data_strides is not C contiguous")
+    if not cnp.PyArray_CHKFLAGS(result, cnp.NPY_C_CONTIGUOUS):
+        raise ValueError(u"result is not C contiguous")
     with nogil:
         for i in range(len_points):
-            _trilinear_interpolation_iso(&ps[i*3],<double *>w,<cnp.npy_intp *>index)
-            rs[i]=0
+            _trilinear_interpolation_iso(
+                &ps[i * 3],
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: indentation

File: dipy/tracking/propspeed.pyx [hunk 3]
```diff
@@ -89,73 +88,104 @@ def ndarray_offset(cnp.ndarray[cnp.npy_intp, ndim=1] indices, \
...
+                value = ds[off]
+                rs[i] += weight * value
     return
 
-cdef  void _trilinear_interpolation_iso(double *X, double *W, cnp.npy_intp *IN) nogil:
 
-    ''' interpolate in 3d volumes given point X
+cdef void _trilinear_interpolation_iso(
+    double *X,
+    double *W,
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: same as above

File: dipy/tracking/propspeed.pyx [hunk 4]
```diff
@@ -174,279 +205,319 @@ cdef  void _trilinear_interpolation_iso(double *X, double *W, cnp.npy_intp *IN)
...
-        tracking
-
-        qa : array, shape(Np,), float, quantitative anisotropy matrix,
-        where Np the number of peaks, found using self.Np
-
-        ind : array, shape(Np,), float, index of the track orientation
 
-        odf_vertices : array, shape(N,3), float, sampling directions on the sphere
+cdef cnp.npy_intp _nearest_direction(
+    double* dx,
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: same as above.

File: dipy/tracking/propspeed.pyx [hunk 5]
```diff
@@ -174,279 +205,319 @@ cdef  void _trilinear_interpolation_iso(double *X, double *W, cnp.npy_intp *IN)
...
 
 
 @cython.cdivision(True)
-cdef cnp.npy_intp _propagation_direction(double *point,double* dx,double* qa,\
-                                double *ind, double *odf_vertices,\
-                                double qa_thr, double ang_thr,\
-                                cnp.npy_intp *qa_shape,cnp.npy_intp* strides,\
-                                double *direction,double total_weight) nogil:
+cdef cnp.npy_intp _propagation_direction(
+    double *point,
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: same as above

File: dipy/tracking/propspeed.pyx [hunk 6]
```diff
@@ -174,279 +205,319 @@ cdef  void _trilinear_interpolation_iso(double *X, double *W, cnp.npy_intp *IN)
...
+    cnp.npy_intp* strides,
+    double *direction,
+    double total_weight) nogil:
     cdef:
-        double total_w=0 #total weighting useful for interpolation
-        double delta=0 #store delta function (stopping function) result
-        double new_direction[3] #new propagation direction
-        double w[8],qa_tmp[PEAK_NO],ind_tmp[PEAK_NO]
-        cnp.npy_intp index[24],i,j,m,xyz[4]
+        double total_w=0 # total weighting useful for interpolation
```
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: pep 8

File: dipy/tracking/propspeed.pyx (hunk 1 above)
Comment (Garyfallidis) [inline dipy/tracking/propspeed.pyx]: Yeah, I see that data_strides goes a bit out. But look below too. There are other places that this is not the case.

**Comment by Garyfallidis** [comment]
Okay, apart from a few pep8 problems. I think this is ready. I tried also with some real data and the tracking looks good. Matthew here is using cnp.PyArray_DATA to access numpy arrays. We may want to use this trick in other places too.

**Comment by arokem** [comment]
Writing up a new PR to address the PEP8 issues. Should be up in just a
minute.

**Comment by matthew-brett** [comment]
Closing in favor of #544

**Comment by matthew-brett** [comment]
Guys - OK to upload the patched wheel to pypi?

**Comment by arokem** [comment]
Seems fine to me.

**Pull request closed without merging**